src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -13,36 +13,64 @@ namespace System.Net.Http
     /// <summary>
     /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
     /// </summary>
-    internal sealed class DiagnosticsHandler : DelegatingHandler
+    internal sealed class DiagnosticsHandler : HttpMessageHandlerStage
     {
         private static readonly DiagnosticListener s_diagnosticListener =
                 new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
 
-        /// <summary>
-        /// DiagnosticHandler constructor
-        /// </summary>
-        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler.
-        /// Note that DiagnosticHandler is the latest in the pipeline </param>
-        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly TextMapPropagator _propagator;
+        private readonly HeaderDescriptor[]? _propagatorFields;
+
+        public DiagnosticsHandler(HttpMessageHandler innerHandler, TextMapPropagator? propagator = null, bool autoRedirect = false)
         {
+            Debug.Assert(IsGloballyEnabled());
+
+            _innerHandler = innerHandler;
+            _propagator = propagator ?? TextMapPropagator.Current;
+
+            if (autoRedirect)
+            {
+                // Prepare HeaderDescriptors for fields we need to clear when following redirects
+                // If a PassThroughPropagator is used, we don't have to clear the headers as they will always be the same
```

Comment:
Why is this a valid assumption?

Reply:
A `PassThroughPropagator` will ignore the current Activity and walk the parents to the root `Activity`.

On a redirect, the root `Activity` will be the same for the initial request as for any automatic redirect following it. The headers we inject will therefore also be the same.